File: .tmp_py4013/batch_04_sources/2513_minimize_the_maximum_of_two_arrays.js
// LeetCode 2513 - Minimize the Maximum of Two Arrays
// https://leetcode.com/problems/minimize-the-maximum-of-two-arrays/

/**
 * @param {number} divisor1
 * @param {number} divisor2
 * @param {number} uniqueCnt1
 * @param {number} uniqueCnt2
 * @return {number}
 */
var minimizeSet = function(divisor1, divisor2, uniqueCnt1, uniqueCnt2) {
    const gcd = (a, b) => {
        while (b !== 0) {
            const t = a % b;
            a = b;
            b = t;
        }
        return a;
    };
    const lcm = Math.floor(divisor1 / gcd(divisor1, divisor2)) * divisor2;
    const ok = (x) => {
        const a = x - Math.floor(x / divisor1);
        const b = x - Math.floor(x / divisor2);
        const both = x - Math.floor(x / lcm);
        return a >= uniqueCnt1 && b >= uniqueCnt2 && both >= uniqueCnt1 + uniqueCnt2;
    };
    let lo = 1, hi = 2 ** 62;
    while (lo < hi) {
        const mid = Math.floor((lo + hi) / 2);
        if (ok(mid)) hi = mid;
        else lo = mid + 1;
    }
    return lo;
};

File: .tmp_rs4013/rust_check_batch08/2513_minimize_the_maximum_of_two_arrays.rs
struct Solution;
// LeetCode 2513 - Minimize the Maximum of Two Arrays
// https://leetcode.com/problems/minimize-the-maximum-of-two-arrays/

impl Solution {
    pub fn minimize_set(divisor1: i32, divisor2: i32, unique_cnt1: i32, unique_cnt2: i32) -> i32 {
        fn gcd(mut x: i64, mut y: i64) -> i64 {
            while y != 0 {
                let t = x % y;
                x = y;
                y = t;
            }
            x
        }
        let lcm = divisor1 as i64 / gcd(divisor1 as i64, divisor2 as i64) * divisor2 as i64;
        let ok = |x: i64| {
            let a = x - x / divisor1 as i64;
            let b = x - x / divisor2 as i64;
            let both = x - x / lcm;
            a >= unique_cnt1 as i64 && b >= unique_cnt2 as i64 && both >= unique_cnt1 as i64 + unique_cnt2 as i64
        };
        let mut lo = 1i64;
        let mut hi = 1i64 << 62;
        while lo < hi {
            let mid = (lo + hi) / 2;
            if ok(mid) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        lo as i32
    }
}

fn main() {}
